ms2pip/predict_xgboost.py

```diff
@@ -0,0 +1,164 @@
+        # Get predictions from XGBoost model
+        bst = xgb.Booster({"nthread": num_cpu})
+        bst.load_model(os.path.join(os.path.expanduser("~"), ".ms2pip", model_file))
+        preds = bst.predict(feature_vectors)
+
+        # Reshape into arrays for each peptide
+        preds = _split_list_by_lengths(preds, num_ions)
+        if ion_type in ["x", "y", "z"]:
+            preds = [np.array(x[::-1], dtype=np.float32) for x in preds]
+        elif ion_type in ["a", "b", "c"]:
+            preds = [np.array(x, dtype=np.float32) for x in preds]
+        else:
+            raise ValueError(f"Unsupported ion_type: {ion_type}")
+        preds_list.append(preds)
+
+    predictions = [list(t) for t in zip(*preds_list)]
+
+    # List of objects with `get` method is expected, use spoofer class
+    return [
+        _MultiprocessingResultSpoofer(
+            (mzs, predictions, None, peptide_lengths, charges, spec_ids)
+        )
+    ]
+
+
+def _get_ms2pip_data_for_xgb(peprec, model_params, ptm_ids):
+    """Get feature vectors and mz values for all peptides in self.data."""
+    peaks_version = model_params["peaks_version"]
+
+    vector_list = []
+    mz_list = []
+    for row in peprec.to_dict(orient="records"):
+
+        peptide = np.array(
+            [0] + [AMINO_ACID_IDS[x] for x in row["peptide"].replace("L", "I")] + [0],
+            dtype=np.uint16,
+        )
+        modpeptide = apply_mods(peptide, row["modifications"], ptm_ids)
+        charge = row["charge"]
+
+        vector_list.append(
+            np.array(
+                ms2pip_pyx.get_vector(peptide, modpeptide, charge), dtype=np.uint16
+            )
+        )
+        mzs = ms2pip_pyx.get_mzs(modpeptide, peaks_version)
+        mz_list.append([np.array(m, dtype=np.float32) for m in mzs])
+
+    feature_vectors = np.vstack(vector_list)
+
+    return feature_vectors, mz_list
+
+
+def _split_list_by_lengths(list_in, lengths):
+    list_in = iter(list_in)
+    return [list(islice(list_in, elem)) for elem in lengths]
+
+
+class _MultiprocessingResultSpoofer:
+    """Spoof result structure of multiprocessing, for direct XGB predictions."""
+
+    def __init__(self, contents):
+        self.contents = contents
+
+    def get(self):
+        return self.contents
+
+
+def apply_mods(peptide, mods, PTMmap):
+    """
+    Takes a peptide sequence and a set of modifications. Returns the modified
+    version of the peptide sequence, c- and n-term modifications. This modified
+    version are hard coded in ms2pipfeatures_c.c for now.
+    """
+    modpeptide = np.array(peptide[:], dtype=np.uint16)
+
+    if mods != "-":
+        l = mods.split("|")
+        if len(l) % 2 != 0:
+            raise InvalidModificationFormattingError(mods)
+        for i in range(0, len(l), 2):
+            tl = l[i + 1]
+            if tl in PTMmap:
+                modpeptide[int(l[i])] = PTMmap[tl]
+            else:
+                raise UnknownModificationError(tl)
+
+    return modpeptide
```

Comment:
Why is this duplicated from ms2pipC?

Reply:
ImportError: cannot import name 'apply_mods' from partially initialized module 'ms2pip.ms2pipC' (most likely due to a circular import) (/home/compomics/ms2pip_c/ms2pip/ms2pipC.py)

Reply:
We should move this to a separate module, to avoid the circular import

Reply:
TODO in the next PR